workloads/imagenet/imagenet_pytorch/workload.py

```diff
@@ -0,0 +1,235 @@
+    for (images, labels) in self._eval_ds:
+      images, labels = self.preprocess_for_eval(images, labels, None, None)
+      logits, _ = self.model_fn(
+          params,
+          images,
+          model_state,
+          spec.ForwardPassMode.EVAL,
+          model_rng,
+          update_batch_norm=False)
+      # TODO(znado): add additional eval metrics?
+      batch_metrics = self._eval_metric(logits, labels)
+      total_metrics = {
+          k: v + batch_metrics[k] for k, v in total_metrics.items()
+      }
+      num_batches += 1
+    # Note (runame): the returned result is not exact because the last batch
+    # has a smaller batch size
+    return {k: float(v / num_batches) for k, v in total_metrics.items()}
+
+  def _build_dataset(self,
+      data_rng: spec.RandomState,
+      split: str,
+      data_dir: str,
+      batch_size: int):
+
+    is_train = (split == "train")
+
+    normalize = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize(
+            mean=[i/255 for i in self.train_mean], 
+            std=[i/255 for i in self.train_stddev])
+    ])
+    transform_config = {
+        "train": transforms.Compose([
+            transforms.RandomResizedCrop(
+                self.center_crop_size,
+                scale=self.scale_ratio_range,
+                ratio=self.aspect_ratio_range),
+            transforms.RandomHorizontalFlip(),
+            normalize]),
+        "test": transforms.Compose([
+            transforms.Resize(self.resize_size),
+            transforms.CenterCrop(self.center_crop_size),
+            normalize])
+    }
+
+    folder = {'train': 'train', 'test': 'val'}
+
+    dataset = ImageFolder(
+        os.path.join(data_dir, folder[split]),
+        transform=transform_config[split])
+
+    dataloader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=is_train,
+        num_workers=5,
+        pin_memory=True,
+        # Note (runame): not sure why we do this?
+        drop_last=is_train,
+        # Note (runame): if we want to use fast_collate,
+        # we need to change some things
+        #collate_fn=fast_collate
+    )
+
+    if is_train:
+      dataloader = cycle(dataloader)
+
+    return dataloader
+
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: spec.Tensor,
+      selected_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor,
+      rng: spec.RandomState) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    del rng
+    return self.preprocess_for_eval(
+        selected_raw_input_batch, selected_label_batch, None, None)
+
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: spec.Tensor,
+      raw_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    return (raw_input_batch.float().to(DEVICE), raw_label_batch.to(DEVICE))
```

Comment:
Currently, this happens in the submission files because the transfer to the `DEVICE` can only happen after the batch was retrieved by the `data_selection()` function, which is also part of the submission. One option would be to write a wrapper class for the `input_queue`, which is also an iterator and transfers the batches to the `DEVICE` when `next()` is called on it. This wrapper could be applied in each PyTorch workload where this behavior is desired.

I think the most important thing is to be consistent across all workloads, i.e. there should be very clear guidelines which functions have to be called in the submissions of the users and what exactly their responsibilities are. Also, I think it might make sense to give the user direct control over the device to use and (in the GPU case) the number of GPUs to use -- at least in "debug mode".

Reply:
To me it sounds like this should be a separate issue/PR.

@runame, if you agree could you formalize the above points in an issue and we could re-visit this in the near future.

Reply:
I have created an issue for this (#50). In it I also mentioned the probably simplest solution for transferring the batches to the `DEVICE`: just call `input_batch.to(DEVICE)` in `workload.model_fn()`. In this way, the batches are always on the appropriate device for each forward pass during training and evaluation. In any case, the issue is still needed to decide on how and where the global `DEVICE` should be set for all workloads.

Reply:
Actually should I remove the `preprocess_for_{train|eval}()` functions and just implement the device transfer as I described above for all workloads where this applies (definitely MNIST/ImageNet PyTorch) in this PR already? Then only the global device setting would be an open issue. I guess it depends on wether you think there are any downsides to my described approach, I think for regular training and evaluation it should be fine though.

Reply:
Since you have the current approach implemented already, I think the best approach would be to finalize this PR and have a separate PR for fixing #50 for all workloads.
This keeps it easy to trace back our changes.